File: hill_cipher.java
import java.util.Scanner;

public class hill_cipher {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Hill Cipher");
        while (true) {
            System.out.print("Choose an option (E)ncrypt, (D)ecrypt, (Q)uit: ");
            String choice = scanner.nextLine().trim().toUpperCase();

            if (choice.equals("Q")) {
                break;
            } else if (choice.equals("E") || choice.equals("D")) {
                System.out.print("Enter the text (uppercase letters only): ");
                String text = scanner.nextLine().trim().toUpperCase();

                System.out.print("Enter the key (9 characters - 3x3 matrix, no spaces): ");
                String key = scanner.nextLine().trim().toUpperCase();

                if (isValidInput(text, key)) {
                    if (choice.equals("E")) {
                        String encryptedText = encrypt(text, key);
                        System.out.println("Encrypted text: " + encryptedText);
                    } else {
                        String decryptedText = decrypt(text, key);
                        System.out.println("Decrypted text: " + decryptedText);
                    }
                } else {
                    System.out.println("Invalid input. Text and key should be uppercase letters.");
                }
            } else {
                System.out.println("Invalid option. Please choose 'E', 'D', or 'Q'.");
            }
        }

        scanner.close();
    }

    public static boolean isValidInput(String text, String key) {
        return text.matches("^[A-Z]+$") && key.matches("^[A-Z]{9}$");
    }

    public static String encrypt(String text, String key) {
        int[][] keyMatrix = new int[3][3];
        int k = 0;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                keyMatrix[i][j] = (key.charAt(k++) % 65);
            }
        }

        StringBuilder ciphertext = new StringBuilder();
        for (int i = 0; i < text.length(); i += 3) {
            String block = text.substring(i, i + 3);
            int[][] blockVector = new int[3][1];
            for (int j = 0; j < 3; j++) {
                blockVector[j][0] = (block.charAt(j) % 65);
            }
            int[][] cipherMatrix = new int[3][1];

            for (int j = 0; j < 3; j++) {
                cipherMatrix[j][0] = 0;
                for (int x = 0; x < 3; x++) {
                    cipherMatrix[j][0] += keyMatrix[j][x] * blockVector[x][0];
                }
                cipherMatrix[j][0] = cipherMatrix[j][0] % 26;
            }

            for (int j = 0; j < 3; j++) {
                ciphertext.append((char) (cipherMatrix[j][0] + 65));
            }
        }

        return ciphertext.toString();
    }

    public static String decrypt(String text, String key) {
        int[][] keyMatrix = new int[3][3];
        int k = 0;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                keyMatrix[i][j] = (key.charAt(k++) % 65);
            }
        }

        int[][] inverseKeyMatrix = inverseMatrix(keyMatrix);

        StringBuilder plaintext = new StringBuilder();
        for (int i = 0; i < text.length(); i += 3) {
            String block = text.substring(i, i + 3);
            int[][] blockVector = new int[3][1];
            for (int j = 0; j < 3; j++) {
                blockVector[j][0] = (block.charAt(j) % 65);
            }
            int[][] plainMatrix = new int[3][1];

            for (int j = 0; j < 3; j++) {
                plainMatrix[j][0] = 0;
                for (int x = 0; x < 3; x++) {
                    plainMatrix[j][0] += inverseKeyMatrix[j][x] * blockVector[x][0];
                }
                plainMatrix[j][0] = (plainMatrix[j][0] + 26) % 26; // Ensure positive value
            }

            for (int j = 0; j < 3; j++) {
                plaintext.append((char) (plainMatrix[j][0] + 65));
            }
        }

        return plaintext.toString();
    }

    public static int[][] inverseMatrix(int[][] matrix) {
        int det = (matrix[0][0] * matrix[1][1] - matrix[0][1] * matrix[1][0]) % 26;
        int detInverse = -1;
        for (int i = 1; i < 26; i++) {
            if ((det * i) % 26 == 1) {
                detInverse = i;
                break;
            }
        }

        int[][] inverse = new int[3][3];
        inverse[0][0] = (matrix[1][1] * detInverse) % 26;
        inverse[0][1] = (-matrix[0][1] * detInverse + 26) % 26;
        inverse[1][0] = (-matrix[1][0] * detInverse + 26) % 26;
        inverse[1][1] = (matrix[0][0] * detInverse) % 26;

        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                inverse[i][j] = (inverse[i][j] + 26) % 26; // Ensure positive values
            }
        }

        return inverse;
    }
}

File: Hill_Cipher.py
def matrix_multiply(matrix1, matrix2):
    result = []
    for i in range(len(matrix1)):
        row = []
        for j in range(len(matrix2[0])):
            product = 0
            for k in range(len(matrix2)):
                product += matrix1[i][k] * matrix2[k][j]
            row.append(product % 26)
        result.append(row)
    return result

def validate_key(key_matrix):
    determinant = (key_matrix[0][0] * key_matrix[1][1]) - (key_matrix[0][1] * key_matrix[1][0])
    determinant %= 26
    if determinant == 0:
        return False
    return True

def hill_cipher(text, key_matrix, mode='encrypt'):
    text = text.upper().replace(" ", "")
    
    if not validate_key(key_matrix):
        raise ValueError("Invalid key. Make sure it's a non-singular integer matrix.")

    n = len(key_matrix)
    if len(text) % n != 0:
        raise ValueError(f"The length of the text must be a multiple of {n}.")

    result = []
    alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

    if mode == 'decrypt':
        determinant = (key_matrix[0][0] * key_matrix[1][1]) - (key_matrix[0][1] * key_matrix[1][0])
        determinant %= 26
        inverse_det = None

        for i in range(26):
            if (determinant * i) % 26 == 1:
                inverse_det = i
                break
        
        if inverse_det is None:
            raise ValueError("The key is not invertible in the modulo 26 field.")

        key_matrix[0][0], key_matrix[1][1] = key_matrix[1][1], key_matrix[0][0]
        key_matrix[0][1] *= -1
        key_matrix[1][0] *= -1

        for i in range(n):
            for j in range(n):
                key_matrix[i][j] = (key_matrix[i][j] * inverse_det) % 26

    for i in range(0, len(text), n):
        block = text[i:i + n]
        block_vector = [alphabet.index(letter) for letter in block]
        
        if mode == 'encrypt':
            result_matrix = matrix_multiply([block_vector], key_matrix)
        elif mode == 'decrypt':
            result_matrix = matrix_multiply([block_vector], key_matrix)

        result += [alphabet[val] for val in result_matrix[0]] #type: ignore 

    return ''.join(result)

def main():
    print("Hill Cipher")
    while True:
        choice = input("Choose an option (E)ncrypt, (D)ecrypt, (Q)uit: ").strip().upper()
        if choice == "Q":
            break
        elif choice == "E" or choice == "D":
            text = input("Enter the text: ").strip()
            key_str = input("Enter the key matrix as space-separated integers (e.g., '3 2 2 5' for a 2x2 matrix): ").strip()
            key = [int(x) for x in key_str.split()]
            key_matrix = [[key[0], key[1]], [key[2], key[3]]]

            if choice == "E":
                result = hill_cipher(text, key_matrix, 'encrypt')
                print("Encrypted text:", result)
            else:
                result = hill_cipher(text, key_matrix, 'decrypt')
                print("Decrypted text:", result)
        else:
            print("Invalid option. Please choose 'E', 'D', or 'Q'.")

if __name__ == "__main__":
    main()
